design/crd-status.md

```diff
@@ -0,0 +1,312 @@
+# Exposing the status with CRDs
+
+## Summary
+
+The purpose of this enhancement is to expose useful information to perform troubleshooting
+via CRDs.
+
+## Motivation
+
+Exposing part of the internal state was not possible until the introduction of CRD
+based configuration, and despite part of the status is exposed via Prometheus
+metrics, troubleshooting MetalLB often requires inspecting the logs of the various
+controllers, and it's not always easy to understand why a service is not working.
+
+For example, inspecting logs is required when:
+
+* an invalid configuration was applied
+* the session with a BGP peer is not established
+* a service was not announced to a BGP peer
+
+### Goals
+
+* Expose useful information to ease the debug
+
+A high level list of informations we want to retrieve is:
+
+* configuration status (valid/invalid and reason)
+* number of used / available IPs per IPAddressPool
```

Comment:
that (especially the unassigned part) can be pretty big in case of v6, so I'd rather avoid.

Reply:
So here are a number of IP statistics, Do not consider showing the specific IP list that has been used or unused?

Reply:
not sure I understood the question. I thought you were talking about adding the full list of ips, one by one

Reply:
Maybe like this: 

```yaml
spec:
  addresses:
  - 10.6.168.140-10.6.168.150
  autoAssign: true
  avoidBuggyIPs: true
status:
  allocatedIPs:
    10.6.168.140:
       name: svc1
       namespace: default
    10.6.168.141:
       name: svc2
       namespace: default
```

Or :

```yaml
spec:
  addresses:
  - 10.6.168.140-10.6.168.150
  autoAssign: true
  avoidBuggyIPs: true
status:
  unallocatedIPs:
  - 10.6.168.140
  - 10.6.168.141
```

Reply:
yeah the problem is, you might get a big list of unallocated in case of v6, plus I am not really sure it's useful. If you allow a service to pick an ip from a pool, it means you don't care which ip it is going to take.

Reply:
This feature is really useful when we assign IPs to Services via annotations: `metallb.universe.tf/loadBalancerIPs`.  When creating the service we can know which IPs are available for assignment which cannot.

Reply:
I think it's useful to show the allocated IPs, then the user can infer which ones are still available

Reply:
I am still not entirely sold on this. I'll add a section with "open items" in the bottom of the doc, which we can use as a starting point for the discussion.